server/src/main/scala/io/delta/standalone/internal/DeltaSharingCDCReader.scala

```diff
@@ -0,0 +1,281 @@
+   * or the version itself.
+   */
+  private def getVersionForCDC(
+      options: Map[String, String],
+      versionKey: String,
+      timestampKey: String,
+      latestVersion: Long
+  ): Option[Long] = {
+    if (options.contains(versionKey)) {
+      Some(options(versionKey).toLong)
+    } else if (options.contains(timestampKey)) {
+      val tsStr = options(timestampKey)
+      if (timestampKey == DeltaDataSource.CDF_START_TIMESTAMP_KEY) {
+        // For the starting timestamp we need to find a version after the provided timestamp
+        // we can use the same semantics as streaming.
+        val ts = getTimestamp("startingTimestamp", tsStr)
+        val resolvedVersion = getStartingVersionFromTimestamp(ts, latestVersion)
+        Some(resolvedVersion)
+      } else {
+        // For ending timestamp the version should be before the provided timestamp.
+        val ts = getTimestamp("endingTimestamp", tsStr)
+        Some(history.getActiveCommitAtTime(ts).version)
+      }
+    } else {
+      None
+    }
+  }
+
+  /**
+   * Replay Delta transaction logs and return cdf files
+   *
+   * @param cdfOptions to indicate the starting and ending version of the change data feed.
+   * @param latestVersion the latest version of the delta table, which is used to validate the
+   *                      starting and ending versions, and may be used as default ending version.
+   */
+  def queryCDF(cdfOptions: Map[String, String], latestVersion: Long): (
+    Seq[CDCDataSpec[AddCDCFile]],
+    Seq[CDCDataSpec[AddFile]],
+    Seq[CDCDataSpec[RemoveFile]]
+  ) = {
+    val (start, end) = getCDCVersions(cdfOptions, latestVersion)
+    val changes = deltaLog.getChanges(start, false).asScala.takeWhile(_.getVersion <= end)
+
+    // Correct timestamp values are only available through DeltaHistoryManager.getCommits(). Commit
+    // info timestamps are wrong, and file modification times are wrong because they need to be
+    // monotonized first. This just performs a list (we don't read the contents of the files in
+    // getCommits()) so it's not a big deal.
+    val timestampsByVersion: Map[Long, Timestamp] = {
+      val commits = DeltaSharingHistoryManager.getCommitsSafe(
+        deltaLog.store,
+        deltaLog.logPath,
+        start,
+        end + 1,
+        conf
+      )
+
+      // Note that the timestamps come from filesystem modification timestamps, so they're
+      // milliseconds since epoch and we don't need to deal with timezones.
+      commits.map(f => (f.version -> new Timestamp(f.timestamp))).toMap
+    }
+
+    val changeFiles = ListBuffer[CDCDataSpec[AddCDCFile]]()
+    val addFiles = ListBuffer[CDCDataSpec[AddFile]]()
+    val removeFiles = ListBuffer[CDCDataSpec[RemoveFile]]()
+
+    if (!isCDCEnabledOnTable(deltaLog.getSnapshotForVersionAsOf(start).metadataScala)) {
+      throw DeltaCDFErrors.changeDataNotRecordedException(start, start, end)
+    }
```

Comment:
Should this be moved earlier in the flow (right after we know the start version) ?

Reply:
good point, updated.